pgjdbc/src/test/java/org/postgresql/jdbc/ResourceLockTest.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2004, PostgreSQL Global Development Group
+ * See the LICENSE file in the project root for more information.
+ */
+
+package org.postgresql.jdbc;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.LockSupport;
+
+class ResourceLockTest {
+
+  @Test
+  void testObtainClose() {
+    final ResourceLock lock = new ResourceLock();
+
+    assertFalse(lock.isLocked());
+    assertFalse(lock.isHeldByCurrentThread());
+
+    try (ResourceLock ignore = lock.obtain()) {
+      assertTrue(lock.isLocked());
+      assertTrue(lock.isHeldByCurrentThread());
+    }
+
+    assertFalse(lock.isLocked());
+    assertFalse(lock.isHeldByCurrentThread());
+  }
+
+  @Test
+  void testObtainWhenMultiThreadedExpectLinearExecution() throws InterruptedException, ExecutionException {
+    CallWithResourceLock callWithResourceLock = new CallWithResourceLock();
+
+    int levelOfConcurrency = 5;
+
+    ExecutorService executorService = Executors.newFixedThreadPool(levelOfConcurrency);
+    try {
+      List<Callable<CounterPair>> callables = new ArrayList<>();
+      for (int i = 0; i < levelOfConcurrency; i++) {
+        callables.add(callWithResourceLock::invoke);
+      }
+
+      // expect linear execution
+      List<Future<CounterPair>> results = executorService.invokeAll(callables);
+
+      Set<Integer> preLockSet = new HashSet<>();
+      Set<Integer> postLockSet = new HashSet<>();
+      for (Future<CounterPair> result : results) {
+        CounterPair entry = result.get();
+        preLockSet.add(entry.preLock);
+        postLockSet.add(entry.postLock);
+      }
+
+      assertEquals(levelOfConcurrency, postLockSet.size()); // linear execution inside resource lock block
+      assertEquals(1, preLockSet.size()); // all threads called invoke before any finish
```

Comment:
It is nice you put comments here, however, please consider using assertion messages rather than code comments. Then the messages would be printed in case of failure, especially in CI logs.

```suggestion
      assertEquals(levelOfConcurrency, postLockSet.size(), "postLockSet.size() should be equal to levelOfConcurrency since we expect that <insert more clarification here> ");
      assertEquals(1, preLockSet.size(), "preLockSet.size(). All threads called should observe a single initial value only...."); 
```

Reply:
I would suggest sorting the resulting sets (e.g. via `TreeSet`) and comparing the contents instead of just comparing the count.

I guess the expected values are always 1, 2, 3, 4, 5, so it would be easier to understand why they all should be different, and it would be eaiser to spot the difference should the test fail.

Reply:
Hmmm. We don't have AssertJ (which is great for collections) but I'm not sure of the value on asserting collection contents here in that we are effectively asserting that all callables were executed (hence postLockSet.size() equals levelOfConcurrency).

ie. All 5 (levelOfConcurrency) are executed and only 1 wins the race to obtain the lock (we don't care which one)

Hmmm.

Reply:
Suppose concurrency is 5, and the .size yields 4. Which information does it provide you?
What if the test fails only in CI?

At the same time, if you collect all the outcomes to ArrayList, then the failure message would include much more information for the analysis.  [1,2,3,4] and [1,2,2,3,4] have both set.size()==4  yet they represent different failure cases (missing value vs duplicated value).

PS We need to bring AssertJ indeed, and I thought of using  OpenRewrite to migrate to assertj, however, OpenRewrite fails to parse type annotations: https://github.com/openrewrite/rewrite/issues/2443, https://github.com/openrewrite/rewrite/issues/2444

PPS AssertJ has issues with collections: https://github.com/assertj/assertj/issues/2671 ;)

Reply:
> Suppose concurrency is 5, and the .size yields 4. Which information does it provide you?

That says that one of the callables didn't complete.

> What if the test fails only in CI?

Yeah, we are kind of testing JDK ReentrantLock. Or more specifically as I see it we want to assert:
A) When `obtain()` is called that calls the underlying jdk `lock.lock()`
B) When `close()` is called by try-with-resources the underlying jdk `lock.unlock()` is called
... everything else kind of puts this in the "we are testing the jdk functionality"

As I see it the main thing we are asserting/checking here is that when we have concurrent threads trying to obtain the lock concurrently, only 1 thread wins that race. We could do something like when each thread actually executes keep the System.nanoTime() of it's start and end ... and assert that those time ranges do not overlap (with 5 concurrent requests we get linear execution of those 5 tasks and their execution times don't overlap).

We could do that ... it just feels a bit like we are testing jdk functionality.  Hmmm.